sqlite/ext/comdb2/systablepermissions.c

```diff
@@ -0,0 +1,291 @@
+
+    permissions_cursor *pCur = sqlite3_malloc(sizeof(*pCur));
+    if (pCur == 0)
+        return SQLITE_NOMEM;
+    tran_type *trans = curtran_gettran();
+    if (!trans) {
+        logmsg(LOGMSG_ERROR, "%s cannot create transaction object\n", __func__);
+        return -1;
+    }
+
+    memset(pCur, 0, sizeof(*pCur));
+
+    pCur->ppTables =
+        sqlite3_malloc(sizeof(struct Systable) * vtab->db->aModule.count);
+    HashElem *systbl;
+    for (systbl = sqliteHashFirst(&vtab->db->aModule); systbl;
+         systbl = sqliteHashNext(systbl)) {
+        struct Module *mod = sqliteHashData(systbl);
+        if (!(mod->pModule->access_flag & CDB2_HIDDEN)) {
+            int pos = pCur->nTables;
+            pCur->ppTables[pos].name = strdup(systbl->pKey);
+            pCur->ppTables[pos].flags = mod->pModule->access_flag;
+            pCur->nTables++;
+        }
+    }
+
+    bdb_user_get_all_tran(trans, &pCur->ppUsers, &pCur->nUsers);
+
+    *ppCursor = &pCur->base;
+    curtran_puttran(trans);
+    return SQLITE_OK;
+}
+
+static int permissionsClose(sqlite3_vtab_cursor *cur)
+{
+    int i;
+    permissions_cursor *pCur = (permissions_cursor *)cur;
+    /* Memory for users[] is owned by bdb */
+    for (i = 0; i < pCur->nUsers; ++i) {
+        free(pCur->ppUsers[i]);
+    }
+    free(pCur->ppUsers);
+    for (i = 0; i < pCur->nTables; ++i) {
+        /* Table names were strdup'd */
+        free(pCur->ppTables[i].name);
+    }
+    sqlite3_free(pCur->ppTables);
+    sqlite3_free(pCur);
+    return SQLITE_OK;
+}
+
+static int permissionsNext(sqlite3_vtab_cursor *cur)
+{
+    permissions_cursor *pCur = (permissions_cursor *)cur;
+    ++pCur->iUser;
+    if (pCur->iUser >= pCur->nUsers) {
+        pCur->iUser = 0;
+        ++pCur->iTable;
+    }
+    return SQLITE_OK;
+}
+
+static int permissionsColumn(sqlite3_vtab_cursor *cur, sqlite3_context *ctx,
+                             int i)
+{
+    permissions_cursor *pCur = (permissions_cursor *)cur;
+
+    tran_type *trans = curtran_gettran();
+    if (!trans) {
+        logmsg(LOGMSG_ERROR, "%s cannot create transaction object\n", __func__);
+        return 1;
+    }
+    char *tbl = pCur->ppTables[pCur->iTable].name;
+    int flags = pCur->ppTables[pCur->iTable].flags;
+    char *usr = pCur->ppUsers[pCur->iUser];
+    switch (i) {
+    case STTP_TABLE: {
+        sqlite3_result_text(ctx, tbl, -1, NULL);
+        break;
+    }
+    case STTP_USER: {
+        sqlite3_result_text(ctx, usr, -1, NULL);
+        break;
+    }
+    case STTP_READ:
+    case STTP_WRITE:
+    case STTP_DDL: {
+        int access, err;
+        if (i == STTP_READ) {
+            access = ACCESS_READ;
+        } else if (i == STTP_WRITE) {
+            access = ACCESS_WRITE;
+        } else {
+            access = ACCESS_DDL;
+        }
+        sqlite3_result_text(
+            ctx,
+            YESNO(((access == ACCESS_READ) && (flags & CDB2_ALLOW_ALL)) ||
```

Comment:
I don't think I understand it. Why do we need a special logic here for READ?

Reply:
If the module has `CDB2_ALLOW_ALL` set, that means it's world readable - no need to check the READ permissions further.